katas/content/multi_qubit_measurements/state_modification/Verification.qs

```diff
@@ -37,20 +37,21 @@ namespace Kata.Verification {
                 // Prepare the state to be input to the testImplementation
                 StateInitialize_StateSelection(alpha, qs);
 
-                // operate testImplementation
                 Kata.StateSelection(qs, ind);
-                // reset the first qubit, since its state does not matter
-                Reset(qs[0]);
 
-                // apply adjoint reference operation and check that the result is correct
+                // Apply adjoint of state preparation operation
                 Adjoint StatePrepare_StateSelection(alpha, ind, qs[1]);
 
-                if not CheckAllZero(qs) {
```

Comment:
@tcNickolas , gates on two different qubits always commute. Even measurements. What is the case that didn't work? The exact reason is probably subtly different?

Reply:
Reset is not a gate, it's a measurement, and CheckAllZero is not a gate either. A measurement collapses the state of the system. 

We start in ∣0⟩⊗(a∣0⟩+b∣1⟩) + ∣1⟩⊗(b∣0⟩+a∣1⟩) (not spelling out normalization coefficients here)

Empty solution leaves this state unchanged, so when we apply adjoint of prep to the second qubit, we end up in 
∣0⟩⊗∣0⟩ + ∣1⟩⊗|non-0⟩. CheckAllZero sees the amplitude of the second basis state and returns false, which is correct

If we measure the first qubit before CheckAllZero, the state collapses, we end up either in ∣0⟩⊗∣0⟩ or in ∣1⟩⊗|non-0⟩ with certain probabilities, and if we hit the first scenario, CheckAllZero will return true.

If we hit the first scenario on each loop iteration, the test will pass for the empty solution

Reply:
Thanks for explaining. As far as I understood it, the 'Check' is not a gate (completely true) and therefore putting it after reset makes the whole thing probabilistic. It still verifies solution but is not always correct. While keeping it before reset peaks into state and therefore is non-probabilistic.

Reply:
Yes, doing Reset before check introduces this uncertainty, since it effectively emulates part of the correct solution.